### ISSUE TITLE
Issues with onVisibleChanged and visibleIndex 

### ISSUE THREAD

**Issue by halexiev-hedgeserv** [surveyjs/surveyjs#1091]
### Are you requesting a feature, reporting a bug or ask a question?

Bug

### What is the current behavior?

We have the following scenario:

Multiple root panels with questions and child panels inside them.
First panel contains a question that all other root panels depend on in terms of visibility, i.e. until the question is answered rest of the root panels are not visible.

Whenever a question becomes visible we have custom onVisibleChanged logic that depends on the visibleIndex. 

There are two issues with onVisibleChanged and visibleIndex: 

1. onVisibleChanged not called for first question
2. visibleIndex is -1 even though questions are rendered within the panels

EDIT: Only when 'no' is clicked after 'yes' on first question visibleIndex values are printed correctly in the console but the questions are not visible anymore.

### What is the expected behavior?

1. onVisibleChanged should be called when the survey starts because the first question is in the first root panel which is visible.
2. visibleIndex should not be -1 for the remaining questions when rest of the root panels become available.

### How would you reproduce the current behavior (if this is a bug)?

Check the plunk below and monitor the console. Note that:

1. When you open the plunk and the console there is nothing printed there - i.e. onVisibleChanged not called.
2. When you expand first panel and click 'yes' under q1 rest of the root panels become visible but the printed visibleIndex for the questions inside the panels is -1 which is wrong, i.e. visibleIndex should increase from 0 to 3.

#### Provide the test code and the tested page URL (if applicable)

Tested page URL: https://plnkr.co/edit/f33lfJAVTUd5aZmFgE3k?p=preview

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv Yes, the question.visibleIndex is incorrect in the event.

Thank you,
Andrew

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv After some investigation, I found the reason why visibleIndex is incorrect.
1) Here is the correct code on onVisibleChanged event
```
survey.onVisibleChanged.add(function (survey, options) {
  console.log(options.question.id + ', visibleIndex=' + options.question.visibleIndex);
});
```
2) The event is fired when the question.visible is changed, however at this moment the question.visibleIndex is -1, because the question is still invisible. The reason is because it's panel is invisible yet. In your example, it has the same visibleIf expression + even if you remove it, it would not help, because at first visibleIf expressions are excecuted in panel chilren elements and then on panel itself. The panel/page is invisible if all question inside them are invisible too. 
So it first question.visible changed from false to true, but panel is still invisible, then panel becomes visible and visibleIndex of all questions are updated.

Anyway, what is your scenario, what problem are you trying to solve? Probably you need another event, maybe a new one.

Thank you,
Andrew

**Comment by halexiev-hedgeserv** [comment]
@andrewtelnov I forked the plunk above and implemented our use case - based on the visibleIndex we highlight the even questions across the whole survey no matter how many panels they are nested in.

We are using the visibleIndex because it is what we need in order to calculate if the question is odd or even survey wise - this means that there cannot be two questions next to each other (disregarding the panels they are in) that are highlighted with same background colour.

The plunk below does not use the **onVisibleChanged**, but the **onGetQuestionTitle** event just so you can see our case because the visibleIndex provided there is correct - I still think that visibleIndex in onVisibleChanged event is **incorrect**. 

**In other words - why can't onVisibleChanged visibleIndex behave just like the onGetQuestionTitle visibleIndex?**

https://plnkr.co/edit/xbs3ISPNOlm38Gi2rsjB?p=preview

Thanks,
Hristo

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv I think we have to add odd/even styles to make easy to implement your scenario. I would also recommend to use onAfterRenderQuestion for this task.

Here is what happens with onVisibleChanged event:
1) survey get notification about chaning a value
2) it tells to all it's pages that there is a value change
3) page - a container - tells all its children (containers/panels and questions) that a value has been changed
4) a container/panel tells a question inside it that a value has been changed. Please note, a panel is invisible at the current moment
5) question runs a visibleIf expression and as result changes its visible property from false to true
6) question send notification to survey object that its visible property has been changed
7) survey runs a function to update all questions visible indexes and fires onVisibleChanged event
8) queston.visibleIndex is -1 because question.isVisible property is false bacause its parent/panel is invisible at this moment yet.

onGetQuestionTitle is fired before rendering to let you change the question title to what ever you want. At this time the question visible index is alread set correctly, because the question parent/panel is visible. Otherwise, onGetQuestionTitle would not fired, since the question would not rendered and there is no need to fire onGetQuestionTitle event.

I hope, It is clear now :-)

Thank you,
Andrew

**Comment by halexiev-hedgeserv** [comment]
@andrewtelnov Thank you for the in-depth explanation! 

I do not mean to argue needlessly or anything like that, but there is still something fishy with these events and the visibleIndex. I forked another plunk (below) and here's the output in the console from the events:

10:51:20.737 index.js:177 **onVisibleChanged: sq_105, visible=true, visibleIndex=-1, parentVisible=true**
...
10:51:20.773 index.js:171 **onAfterRenderQuestion: sq_105, visible=true, visibleIndex=-1, parentVisible=true**

So you can see that in both **onVisibleChanged** and **onAfterRenderQuestion** events the question is visible, its parent is visible, but the visibleIndex of the question is still -1 ... just does not make sense :(

We are going to be using the onGetQuestionTitle for the time being because this is what seems to be working for us, but it kind of feels like a workaround.

https://plnkr.co/edit/85bkwJbCS5JjhZTnvy6d?p=preview

Thanks,
Hristo

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv Yes, indeed, parent.isVisible is already true. I will take a look.

Thank you,
Andrew

**Comment by halexiev-hedgeserv** [comment]
Hi @andrewtelnov, any luck with this?

Thanks,
Hristo

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv Sorry, I have not looked at it yet. But I will.

Thank you,
Andrew

**Comment by andrewtelnov** [comment]
@halexiev-hedgeserv It was actually the bug. I has made some optimization before, but it fact there is no need in it. I have removed it.
I have added an unit test and fixed it.

Thank you,
Andrew

**Comment by halexiev-hedgeserv** [comment]
@andrewtelnov Thank you!
